### PR TITLE
Forward recompute2

### PR DESCRIPTION
**Pull request by mapingshuo**
forward recompute back-propagation. This method is able to increase training batch size up to 10x or make current model 5x deeper.

### CODE REVIEW

File: python/paddle/fluid/optimizer.py [hunk 1]
```diff
@@ -2973,6 +2975,254 @@ def minimize(self,
...
+        if self._checkpoints == None:
+            raise ValueError("You should call _set_checkpoints first")
+
+        if framework.in_dygraph_mode():
+            raise NotImplementedError(
+                "DyGraph current does not support recompute")
+
+        self._dtype = loss.dtype
+        program = loss.block.program
+        #if self.debug == True:
```
Comment (guru4elephant) [inline python/paddle/fluid/optimizer.py]: remove these if not needed？

File: python/paddle/fluid/optimizer.py [hunk 2]
```diff
@@ -2973,6 +2975,254 @@ def minimize(self,
...
+        plt.plot(x, y, label="with_recompute")
+        #plt.plot(x2, y2, label="without_recompute") 
+
+        plt.legend(loc='upper right')
+        plt.xlabel("op_idx")
+        plt.ylabel('MB')
+        plt.title('')
+        plt.savefig('memory_anal.png')
+
+        import SimpleHTTPServer
```
Comment (guru4elephant) [inline python/paddle/fluid/optimizer.py]: this code can be used as external code, better not merge

File: paddle/fluid/API.spec [hunk 3]
```diff
@@ -7,6 +7,16 @@ paddle.fluid.Program.global_block (ArgSpec(args=['self'], varargs=None, keywords
 paddle.fluid.Program.list_vars (ArgSpec(args=['self'], varargs=None, keywords=None, defaults=None), ('document', '32c14b0f12baae4b352200fa09b5e789'))
 paddle.fluid.Program.parse_from_string (ArgSpec(args=['binary_str'], varargs=None, keywords=None, defaults=None), ('document', 'b6a7ffb239a30bf2ce58cfaca8d8b8d5'))
 paddle.fluid.Program.to_string (ArgSpec(args=['self', 'throw_on_error', 'with_details'], varargs=None, keywords=None, defaults=(False,)), ('document', '89acca639baf00f3ad08b9d827e81706'))
+paddle.fluid.Variable ('paddle.fluid.framework.Variable', ('document', '1f67115bb22a20f688291e819a5dd6b5'))
```
Comment (guru4elephant) [inline paddle/fluid/API.spec]: Variable can be hidden here.

File: python/paddle/fluid/framework.py [hunk 4]
```diff
@@ -36,6 +36,7 @@
 
 __all__ = [
     'Program',
+    'Variable',
```
Comment (guru4elephant) [inline python/paddle/fluid/framework.py]: remove this if not needed.

File: python/paddle/fluid/optimizer.py [hunk 5]
```diff
@@ -2973,6 +2975,254 @@ def minimize(self,
...
+            place = fluid.CPUPlace()
+            exe = fluid.Executor(place)
+            exe.run(fluid.default_startup_program())
+            step = 10
+
+            for i in range(step):
+                cost_val = exe.run(feed=gen_data(),
+                       program=fluid.default_main_program(),
+                       fetch_list=[cost.name])
+                print("step=%d cost=%f" % (i, cost_val[0]))
```
Comment (guru4elephant) [inline python/paddle/fluid/optimizer.py:3039]: this example is good w.r.t its complete code for training.

File: python/paddle/fluid/optimizer.py [hunk 6]
```diff
@@ -2973,6 +2975,254 @@ def minimize(self,
...
+        PORT = 8233
+
+        Handler = SimpleHTTPServer.SimpleHTTPRequestHandler
+
+        httpd = SocketServer.TCPServer(("", PORT), Handler)
+
+        #print("serving at port", PORT)
+        httpd.serve_forever()
+
+    def analysis_memory_usage(self, block):
```
Comment (guru4elephant) [inline python/paddle/fluid/optimizer.py]: maybe you can make this a helper tool outside fluid python API. Once  this function is robust and exactly right w.r.t. GPU memory, you can make this function officially accessed by users.

File: python/paddle/fluid/optimizer.py [hunk 7]
```diff
@@ -2973,6 +2975,254 @@ def minimize(self,
...
+                no_grad_set,
+                checkpoints=self._checkpoints)
+
+            if grad_clip:
+                # TODO(guru4elephant): should add grad_clip for static graph
+                pass
+        optimize_ops = self._optimizer.apply_optimize(
+            loss, startup_program=startup_program, params_grads=params_grads)
+
+        if self.debug == True:
```
Comment (guru4elephant) [inline python/paddle/fluid/optimizer.py]: again, just remove these. and make this file as your debug tool:)

**Comment by mapingshuo** [comment]
image
image

File: python/paddle/fluid/optimizer.py [hunk 8]
```diff
@@ -2973,6 +2975,108 @@ def minimize(self,
         }
 
 
+class RecomputeOptimizer(Optimizer):
+    """
+    Recompute Optimizer Wrapper
+
+    Normally, a training step contains three sub-steps: first, run forward
+    Oprators to calculate the loss; second, run backward Operators to
```
Comment (hutuxian) [inline python/paddle/fluid/optimizer.py]: Oprators -> Operators

File: python/paddle/fluid/optimizer.py [hunk 9]
```diff
@@ -2973,6 +2975,108 @@ def minimize(self,
...
+    Normally, a training step contains three sub-steps: first, run forward
+    Oprators to calculate the loss; second, run backward Operators to 
+    calculate gradient of the parameters; third, apply optimization method
+    to update the value of the parameters.
+
+    In the forward computation process, all variables that are needed by 
+    backward computation process will be kept in memory. They occupy a great
+    amount of memory when the network becomes very deep.
+
+    Recompute split the network to k segements. In each segment, It will
```
Comment (hutuxian) [inline python/paddle/fluid/optimizer.py]: segements -> segments

File: python/paddle/fluid/optimizer.py [hunk 10]
```diff
@@ -2973,6 +2975,108 @@ def minimize(self,
...
+    Oprators to calculate the loss; second, run backward Operators to 
+    calculate gradient of the parameters; third, apply optimization method
+    to update the value of the parameters.
+
+    In the forward computation process, all variables that are needed by 
+    backward computation process will be kept in memory. They occupy a great
+    amount of memory when the network becomes very deep.
+
+    Recompute split the network to k segements. In each segment, It will 
+    recompute the forward Opeartors, before runing backward operators. It is
```
Comment (hutuxian) [inline python/paddle/fluid/optimizer.py]: Opeartors -> Operators
runing -> running

File: python/paddle/fluid/optimizer.py [hunk 11]
```diff
@@ -2973,6 +2975,108 @@ def minimize(self,
...
+    very helpful for saving memory.
+ 
+    The Variables that seperate a network to segments are called as checkpoints,
```
Comment (hutuxian) [inline python/paddle/fluid/optimizer.py]: seperate -> separate

File: python/paddle/fluid/optimizer.py (hunk 10 above)
Comment (mapingshuo) [reply python/paddle/fluid/optimizer.py]: Thx

File: python/paddle/fluid/optimizer.py [hunk 12]
```diff
@@ -2973,6 +2975,108 @@ def minimize(self,
...
+    """
+    Recompute Optimizer Wrapper
+
+    Normally, a training step contains three sub-steps: first, run forward
+    Oprators to calculate the loss; second, run backward Operators to 
+    calculate gradient of the parameters; third, apply optimization method
+    to update the value of the parameters.
+
+    In the forward computation process, all variables that are needed by 
+    backward computation process will be kept in memory. They occupy a great
```
Comment (hutuxian) [inline python/paddle/fluid/optimizer.py]: They -> , which?

**Pull request closed without merging**